sources/api/early-boot-config/src/provider/cdrom.rs

```diff
@@ -0,0 +1,209 @@
+//! The cdrom module implements the `PlatformDataProvider` trait for gathering userdata from a
+//! mounted CDRom.
+
+use super::{PlatformDataProvider, SettingsJson};
+use serde::Deserialize;
+use snafu::{ensure, OptionExt, ResultExt};
+use std::ffi::OsStr;
+use std::fs::{self, File};
+use std::io::BufReader;
+use std::path::Path;
+
+pub(crate) struct CdromDataProvider;
+
+impl CdromDataProvider {
+    // This program expects that the CD-ROM is already mounted.  Mounting happens elsewhere in a
+    // systemd unit file
+    const CD_ROM_MOUNT: &'static str = "/media/cdrom";
+    // A mounted CD-ROM may contain an OVF file or a user-supplied file named `user-data`
+    const USER_DATA_FILENAMES: [&'static str; 5] = [
+        "user-data",
+        "ovf-env.xml",
+        "OVF-ENV.XML",
+        "ovf_env.xml",
+        "OVF_ENV.XML",
+    ];
+
+    /// Given the list of acceptable filenames, ensure only 1 exists and parse
+    /// it for user data
+    fn user_data() -> Result<Option<SettingsJson>> {
+        // Build a list of user data files that exist
+        let mut user_data_files = Self::USER_DATA_FILENAMES
+            .iter()
+            .map(|filename| Path::new(Self::CD_ROM_MOUNT).join(filename))
+            .filter(|file| file.exists());
+
+        // If no files exist, return
+        let user_data_file = match user_data_files.next() {
+            Some(file) => file,
+            None => return Ok(None),
+        };
+
+        // There should only be 1 file, if more exist something is wonky
+        ensure!(
+            user_data_files.next().is_none(),
+            error::UserDataFileCount {
+                location: Self::CD_ROM_MOUNT
+            }
+        );
```

Comment:
nit: for me, all three of these comments are about "what" which is pretty obvious. A "why" comment could be helpful for building a list of the files, since it's not obvious why there wouldn't be a canonical file name.